src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace Microsoft.ML.Runtime.Internal.CpuMath
+{
+    internal static class SseIntrinsics
+    {
+        private static unsafe Vector128<float> Load1(float* src, int* idx)
+        {
+            return Sse.SetScalarVector128(src[idx[0]]);
+        }
+
+        private static unsafe Vector128<float> Load4(float* src, int* idx)
+        {
+            return Sse.SetVector128(src[idx[3]], src[idx[2]], src[idx[1]], src[idx[0]]);
+        }
+
+        private static Vector128<float> Rotate(Vector128<float> x)
+        {
+            return Sse.Shuffle(x, x, 0x39);
+        }
+
+        private static Vector128<float> RotateReverse(Vector128<float> x)
+        {
+            return Sse.Shuffle(x, x, 0x93);
+        }
+
+        // Warning: this operation changes the value of x => do not reuse x
```

Comment:
This comment isn't relevant, `x` is a value type and any modification in this method won't propagate out.

Reply:
Thanks for the comment.  Since this Store4 function has the output type `void`, the input argument `x` should be changed, or the Store4 function would be a no-op.  I copied this comment from `src\Native\Sse.cpp` (https://github.com/dotnet/machinelearning/blob/master/src/Native/CpuMathNative/Sse.cpp#L44).

Reply:
The difference between the native code and the C# code is that the native code is a `#define`, which means that it will directly modify the variable `x` in the context of the containing method.

The C# code is an actual function, and has all the normal safety rules (`x` is a separate local from any `x` in the caller and it is a value type, so any modifications to `x` in `Store4` won't propagate up).

Reply:
Thank you, Tanner!  Since the intended effect of `Store4` is to modify `x`, I will have to change the function a bit.  Should I make `Store4` return `Vector128<float>`, or return `void` with the `out Vector128<float> x` or `ref Vector128<float> x`?  Thank you for your advice in advance!

Reply:
I don't believe the intended effect was to modify `x` and, from a quick scan, I didn't see any code that relied on `x` being modified.

I would assume the comment was added because people were initially calling `Store4`, reusing `x`, and have failures that took some time to debug.